src/sources/gcp_pubsub.rs

```diff
@@ -347,34 +347,33 @@ impl PubsubSource {
         config
     }
 
-    fn request_stream(&self) -> impl Stream<Item = proto::StreamingPullRequest> + 'static {
-        // This data is only allowed in the first request
-        let mut subscription = Some(self.subscription.clone());
-        let mut client_id = Some(CLIENT_ID.clone());
-
-        let ack_ids = Arc::clone(&self.ack_ids);
+    fn request_stream(
+        &self,
+        ack_ids: UnboundedReceiver<Vec<String>>,
+    ) -> impl Stream<Item = proto::StreamingPullRequest> + 'static {
+        let ack_ids = UnboundedReceiverStream::new(ack_ids);
+        let subscription = self.subscription.clone();
+        let client_id = CLIENT_ID.clone();
         let stream_ack_deadline_seconds = self.ack_deadline_seconds;
-        stream::repeat(()).then(move |()| {
-            let ack_ids = Arc::clone(&ack_ids);
-            let subscription = subscription.take().unwrap_or_default();
-            let client_id = client_id.take().unwrap_or_default();
-            async move {
-                let mut ack_ids = ack_ids.lock().await;
-                proto::StreamingPullRequest {
-                    subscription,
-                    client_id,
-                    ack_ids: std::mem::take(ack_ids.as_mut()),
-                    stream_ack_deadline_seconds,
-                    ..Default::default()
-                }
+        stream::once(async move {
+            proto::StreamingPullRequest {
+                subscription,
+                client_id,
+                stream_ack_deadline_seconds,
+                ..Default::default()
             }
         })
+        .chain(ack_ids.map(|ack_ids| proto::StreamingPullRequest {
+            ack_ids,
+            ..Default::default()
```

Comment:
I'm assuming that the initial request which sets the client ID essentially mutates the RPC transaction so that subsequent streaming pull requests continue to be associated with the same client ID?

Reply:
Yes, in fact the client ID and most of those fields are actually not valid in anything but the first request, [according to the docs](https://cloud.google.com/pubsub/docs/reference/rpc/google.pubsub.v1#streamingpullrequest).